File: ui/admin-frontend/src/admin/services/syncStatusService.js
import apiClient from '../utils/apiClient';

class SyncStatusService {
  /**
   * Get sync status summary for all namespaces
   * @returns {Promise<{data: Array, has_pending: boolean, pending_namespaces: Array}>}
   */
  async getSyncStatus() {
    try {
      const response = await apiClient.get('/sync/status');
      return response.data;
    } catch (error) {
      console.error('Error fetching sync status:', error);
      throw new Error(error.response?.data?.error || 'Failed to fetch sync status');
    }
  }

  /**
   * Get detailed sync status for a specific namespace
   * @param {string} namespace - The namespace to get status for
   * @returns {Promise<{namespace: Object, edges: Array}>}
   */
  async getNamespaceSyncStatus(namespace) {
    try {
      const response = await apiClient.get(`/sync/status/${namespace}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching namespace sync status:', error);
      throw new Error(error.response?.data?.error || 'Failed to fetch namespace sync status');
    }
  }

  /**
   * Get sync audit log entries
   * @param {Object} params - Query parameters
   * @param {string} [params.namespace] - Filter by namespace
   * @param {string} [params.edge_id] - Filter by edge ID
   * @param {string} [params.event_type] - Filter by event type
   * @param {number} [params.limit=50] - Maximum number of entries to return
   * @returns {Promise<{data: Array}>}
   */
  async getSyncAuditLog(params = {}) {
    try {
      const response = await apiClient.get('/sync/audit', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching sync audit log:', error);
      throw new Error(error.response?.data?.error || 'Failed to fetch sync audit log');
    }
  }

  /**
   * Check if there are any pending syncs
   * @returns {Promise<boolean>}
   */
  async hasPendingSyncs() {
    try {
      const response = await this.getSyncStatus();
      return response.has_pending || false;
    } catch (error) {
      console.error('Error checking pending syncs:', error);
      return false;
    }
  }

  /**
   * Get sync status color for UI display
   * @param {string} syncStatus - The sync status value
   * @returns {{color: string, label: string}}
   */
  getSyncStatusDisplay(syncStatus) {
    const statusConfig = {
      in_sync: { color: 'success', label: 'Synced' },
      pending: { color: 'warning', label: 'Pending' },
      stale: { color: 'error', label: 'Stale' },
      unknown: { color: 'default', label: 'Unknown' },
    };
    return statusConfig[syncStatus] || statusConfig.unknown;
  }
}

export default new SyncStatusService();

File: services/sync_status_service.go
package services

import (
	"time"

	"github.com/TykTechnologies/midsommar/v2/models"
	"gorm.io/gorm"
)

// SyncStatusService manages configuration sync status between control and edge gateways
type SyncStatusService struct {
	db *gorm.DB
}

// NewSyncStatusService creates a new SyncStatusService
func NewSyncStatusService(db *gorm.DB) *SyncStatusService {
	return &SyncStatusService{db: db}
}

// NamespaceSyncSummary represents sync status for a namespace with edge counts
type NamespaceSyncSummary struct {
	Namespace        string    `json:"namespace"`
	ExpectedChecksum string    `json:"expected_checksum"`
	ConfigVersion    string    `json:"config_version"`
	LastConfigChange time.Time `json:"last_config_change"`
	SyncedCount      int64     `json:"synced_count"`
	PendingCount     int64     `json:"pending_count"`
	StaleCount       int64     `json:"stale_count"`
	UnknownCount     int64     `json:"unknown_count"`
	TotalEdges       int64     `json:"total_edges"`
}

// GetNamespaceSyncSummary returns sync status for all namespaces
func (s *SyncStatusService) GetNamespaceSyncSummary() ([]NamespaceSyncSummary, error) {
	var results []NamespaceSyncSummary

	// Get all namespace sync statuses
	var namespaces []models.NamespaceSyncStatus
	status := &models.NamespaceSyncStatus{}
	namespaces, err := status.GetAll(s.db)
	if err != nil {
		return nil, err
	}

	for _, ns := range namespaces {
		summary := NamespaceSyncSummary{
			Namespace:        ns.Namespace,
			ExpectedChecksum: ns.ExpectedChecksum,
			ConfigVersion:    ns.ConfigVersion,
			LastConfigChange: ns.LastConfigChange,
		}

		// Count edges by sync status for this namespace
		edgeInstance := &models.EdgeInstance{}
		counts, err := edgeInstance.CountEdgesBySyncStatus(s.db, ns.Namespace)
		if err == nil {
			summary.SyncedCount = counts[models.EdgeSyncStatusInSync]
			summary.PendingCount = counts[models.EdgeSyncStatusPending]
			summary.StaleCount = counts[models.EdgeSyncStatusStale]
			summary.UnknownCount = counts[models.EdgeSyncStatusUnknown]
			summary.TotalEdges = summary.SyncedCount + summary.PendingCount + summary.StaleCount + summary.UnknownCount
		}

		results = append(results, summary)
	}

	return results, nil
}

// GetNamespaceSyncStatus returns detailed sync status for a specific namespace
func (s *SyncStatusService) GetNamespaceSyncStatus(namespace string) (*NamespaceSyncSummary, []models.EdgeInstance, error) {
	// Get namespace sync status
	var status models.NamespaceSyncStatus
	if err := status.GetByNamespace(s.db, namespace); err != nil {
		return nil, nil, err
	}

	summary := &NamespaceSyncSummary{
		Namespace:        status.Namespace,
		ExpectedChecksum: status.ExpectedChecksum,
		ConfigVersion:    status.ConfigVersion,
		LastConfigChange: status.LastConfigChange,
	}

	// Get edges in namespace with their sync status
	var edges models.EdgeInstances
	if err := edges.ListEdgesInNamespace(s.db, namespace); err != nil {
		return summary, nil, err
	}

	// Count edges by sync status
	for _, edge := range edges {
		if edge.Status == models.EdgeStatusConnected || edge.Status == models.EdgeStatusRegistered {
			switch edge.SyncStatus {
			case models.EdgeSyncStatusInSync:
				summary.SyncedCount++
			case models.EdgeSyncStatusPending:
				summary.PendingCount++
			case models.EdgeSyncStatusStale:
				summary.StaleCount++
			default:
				summary.UnknownCount++
			}
		}
	}
	summary.TotalEdges = summary.SyncedCount + summary.PendingCount + summary.StaleCount + summary.UnknownCount

	return summary, edges, nil
}

// HasPendingSyncs returns true if any namespace has edges waiting for sync
func (s *SyncStatusService) HasPendingSyncs() (bool, error) {
	var count int64
	err := s.db.Model(&models.EdgeInstance{}).
		Where("sync_status IN ? AND status IN ?",
			[]string{models.EdgeSyncStatusPending, models.EdgeSyncStatusStale},
			[]string{models.EdgeStatusConnected, models.EdgeStatusRegistered}).
		Count(&count).Error
	return count > 0, err
}

// GetPendingNamespaces returns namespaces that have edges pending sync
func (s *SyncStatusService) GetPendingNamespaces() ([]string, error) {
	var namespaces []string
	err := s.db.Model(&models.EdgeInstance{}).
		Distinct("namespace").
		Where("sync_status IN ? AND status IN ?",
			[]string{models.EdgeSyncStatusPending, models.EdgeSyncStatusStale},
			[]string{models.EdgeStatusConnected, models.EdgeStatusRegistered}).
		Pluck("namespace", &namespaces).Error
	return namespaces, err
}

// GetSyncAuditLog returns audit log entries with optional filtering
func (s *SyncStatusService) GetSyncAuditLog(namespace, edgeID, eventType string, limit int) ([]models.SyncAuditLog, error) {
	auditLog := &models.SyncAuditLog{}
	return auditLog.GetFiltered(s.db, namespace, edgeID, eventType, limit)
}

// MarkStaleEdges marks edges that have been out of sync for too long as stale
// staleThreshold is the duration after which pending edges are marked as stale (default 15 minutes)
func (s *SyncStatusService) MarkStaleEdges(staleThreshold time.Duration) error {
	edgeInstance := &models.EdgeInstance{}
	return edgeInstance.MarkStaleEdges(s.db, staleThreshold)
}
